include/nbl/video/IGPUCommandPool.h

```diff
@@ -31,13 +66,219 @@ class IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
         virtual const void* getNativeHandle() const = 0;
 
+        core::CMemoryPool<core::GeneralpurposeAddressAllocator<uint32_t>, core::default_aligned_allocator, false, uint32_t> m_commandSegmentPool;
+
     protected:
         virtual ~IGPUCommandPool() = default;
 
         core::bitflag<E_CREATE_FLAGS> m_flags;
         uint32_t m_familyIx;
 };
 
+class alignas(COMMAND_ALIGNMENT) IGPUCommandPool::ICommand
+{
+public:
+    virtual ~ICommand() {}
+
+    // static void* operator new(std::size_t size) = delete;
+    static void* operator new[](std::size_t size) = delete;
+    // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+    static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+    // static void operator delete  (void* ptr) = delete;
+    static void operator delete[](void* ptr) = delete;
+    static void operator delete  (void* ptr, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::align_val_t al) = delete;
+    static void operator delete  (void* ptr, std::size_t sz) = delete;
+    static void operator delete[](void* ptr, std::size_t sz) = delete;
+    static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+    uint32_t m_size;
+protected:
+    ICommand(uint32_t size) : m_size(size) {}
+};
+
+class alignas(COMMAND_SEGMENT_ALIGNMENT) IGPUCommandPool::CommandSegment
+{
+public:
+    struct Iterator;
+
+    struct params_t
+    {
+        core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+        CommandSegment* m_next;
+    } params;
+
+    static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - sizeof(params_t);
+    alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+    CommandSegment()
+    {
+        params.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, COMMAND_SEGMENT_SIZE);
+        params.m_next = nullptr;
+
+        wipeNextCommandSize();
+    }
+
+    template <typename Cmd, typename... Args>
+    // Cmd* allocate(const Args&... args)
+    Cmd* allocate(Args&&... args)
+    {
+        const uint32_t cmdSize = Cmd::calc_size(args...);
+        const auto address = params.m_commandAllocator.alloc_addr(cmdSize, alignof(Cmd));
+        if (address == decltype(params.m_commandAllocator)::invalid_address)
+            return nullptr;
+
+        wipeNextCommandSize();
+
+        void* cmdMem = m_data + address;
+        return new (cmdMem) Cmd(args...);
+    }
+
+private:
+    void wipeNextCommandSize()
+    {
+        const auto cursor = params.m_commandAllocator.get_allocated_size();
+        const uint32_t wipeSize = offsetof(IGPUCommandPool::ICommand, m_size) + sizeof(IGPUCommandPool::ICommand::m_size);
+        if (cursor + wipeSize < params.m_commandAllocator.get_total_size())
+            memset(m_data + cursor, 0, wipeSize);
```

Comment:
As things are right now, isn't `m_data+cursor` the pointer to first element in `IGPUCommandPool::ICommand` which is a vtable?

Not doing this right now because I'm not sure how should we deal with this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r864263148.

Reply:
Consider the last segment which is half-filled with commands. After the final command there is a region of memory of size `sizeof(ICommand)` bytes but nothing is constructed on that memory so doing something like:
```C++
const bool lastCmd = itr.m_cmd.getSize() == 0; // or even itr.m_cmd.m_size == 0
```
is most likely UB, no?

We should either do a `memcmp` there or construct a dummy `ICommand` object.

Reply:
Resolved.